packages/vgrammar/src/transforms/data/kde.ts

```diff
@@ -0,0 +1,124 @@
+import type { IPointLike } from '@visactor/vutils';
+// eslint-disable-next-line no-duplicate-imports
+import { PointService, array, isNumber } from '@visactor/vutils';
+import type { KDETransformOption } from '../../types';
+
+const defaultBins = 256;
+const defaultAs = ['x', 'kde'];
+const defaultAs2d = ['x', 'y', 'kde'];
+const SQRT2PI = Math.sqrt(Math.PI * 2);
+const SQRT2PI2 = Math.sqrt((Math.PI * 2) ** 2);
+
+// Only use gauss kernel for now
+const gaussKernel = (x: number, dimension: number = 1) => {
+  const sp = dimension === 1 ? SQRT2PI : dimension === 2 ? SQRT2PI2 : Math.sqrt((Math.PI * 2) ** dimension);
+  return Math.exp(-(x ** 2) / 2) / sp;
+};
+
+// A rule-of-thumb bandwidth estimator, referring to: https://en.wikipedia.org/wiki/Kernel_density_estimation#A_rule-of-thumb_bandwidth_estimator
+const ruleOfThumbBandwidth = (data: IPointLike[], dimension: number = 1) => {
+  const n = data.length;
+  const sum = data.reduce((sum, datum) => sum + datum.x, 0);
+  const mean = sum / n;
+  const sd = Math.sqrt(data.reduce((v, datum) => v + (datum.x - mean) ** 2, 0) / n);
+  const sortedData = data.sort((a, b) => a.x - b.x);
+  const q1 = (sortedData[Math.floor(n / 4)].x + sortedData[Math.ceil(n / 4)].x) / 2;
```

Comment:
vutils has a function `quantileSorted`, may be you can use `quantileSorted(sortedData, 0.25)` to replace this line

Reply:
Fixed.